.github/workflows/yocto-builds.yml

```diff
@@ -0,0 +1,65 @@
+# SPDX-FileCopyrightText: Andrei Gherzan <andrei.gherzan@huawei.com>
+#
+# SPDX-License-Identifier: MIT
+
+name: Builds
+
+on:
+  pull_request:
+
+jobs:
+  build:
+    strategy:
+      fail-fast: true
+      matrix:
+        machine:
+          - raspberrypi
+          - raspberrypi0-2w-64
+          - raspberrypi0-2w
+          - raspberrypi0
+          - raspberrypi0-wifi
+          - raspberrypi2
+          - raspberrypi3-64
+          - raspberrypi3
+          - raspberrypi4-64
+          - raspberrypi4
+          - raspberrypi-cm3
+          - raspberrypi-cm
+        image: [core-image-base]
+        dl-dir: [/var/lib/ci/yocto/downloads]
+        sstate-dir: [/var/lib/ci/yocto/sstate]
+        distro: [poky]
+    runs-on: [self-hosted, Linux]
+    name: ${{ matrix.machine }}/${{ matrix.image }}/poky/systemd
+    steps:
+      - name: Checkout the code
+        uses: actions/checkout@v2
+        with:
+          fetch-depth: 0
+      - name: Build a temporary yocto-builder image
+        uses: ./.github/actions/docker-build
```

Comment:
This feels wasteful.

Can you pull from a registry instead?

Reply:
I agree with you that this looks a bit cumbersome and maybe useless. The advantage of having this as such is the ability to modify the docker images (Dockerfiles) along with the extra layer metadata as part of the same PR (repo). I'll test drive this for a while and based on how it behaves I might change to a separate repo for the docker files and use a registry which would make this cleaner but with the disadvantage of having to use multiple repos/PRs.

Reply:
Another solution would be to keep the Dockerfile in this repo, but push to a registry, and then do a pull from that and use that as cache, so you don't have to waste time on rebuilding unless actually needed.

Reply:
I normally restrict the push to registry to only happen for official branches.
Depending on requirements, I bump a version number in the image tag, so old image versions are maintained.

For testing purposes, I either let that require a manual trigger, or push to a "test" image, which is then recycled for each build.

Reply:
Using one test image would have a race condition for multiple PRs. Also, you need to make sure multiple PRs don't use images built for another PR. So it would need per PR images. And even if you do, there is no guarantee that the current PR  image is synced with the current iteration for the code. So you would need to rebuild anyway (with existing layers in the repository so not from scratch as to how I've done it here). The issue though is when you have jobs running for multiple iterations of the same MR due to the fact that stopping jobs in GitHub is not a baked functionality - and it is dealt with usually but an additional pipeline that works on start of PRs.

All these are much easier if you deal with local images where you assume one worker per docker storage. The image cleanup it's a disk optimization because I couldn't find an easy way to run a job on merge (it becomes a push where you don't have the reference of the originating PR). If I figure that out, I could reuse build images locally until I merge when I deal with cleanups.